File: pdu/javascript/nav_msgs/pdu_conv_GridCells.js
import * as PduUtils from '../pdu_utils.js';
import { GridCells } from './pdu_jstype_GridCells.js';
import { Header } from '../std_msgs/pdu_jstype_Header.js';
import { binary_read_recursive_Header, binary_write_recursive_Header } from '../std_msgs/pdu_conv_Header.js';
import { Point } from '../geometry_msgs/pdu_jstype_Point.js';
import { binary_read_recursive_Point, binary_write_recursive_Point } from '../geometry_msgs/pdu_conv_Point.js';


/**
 * Deserializes a binary PDU into a GridCells object.
 * @param {ArrayBuffer} binary_data
 * @returns { GridCells }
 */
export function pduToJs_GridCells(binary_data) {
    const js_obj = new GridCells();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_GridCells(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_GridCells(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: header, type: std_msgs/Header (struct)

    {
        const tmp_obj = new Header();
        binary_read_recursive_Header(meta, binary_data, tmp_obj, base_off + 0);
        js_obj.header = tmp_obj;
    }
    
    // member: cell_width, type: float32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 136, 4);
        js_obj.cell_width = PduUtils.binToValue("float32", bin);
    }
    
    // member: cell_height, type: float32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 140, 4);
        js_obj.cell_height = PduUtils.binToValue("float32", bin);
    }
    
    // member: cells, type: geometry_msgs/Point (struct)

    { // varray
        const array_size = view.getInt32(base_off + 144, littleEndian);
        const offset_from_heap = view.getInt32(base_off + 144 + 4, littleEndian);
        const one_elm_size = 24;
        const array_value = [];
        for (let i = 0; i < array_size; i++) {
            const tmp_obj = new Point();
            const item_offset = meta.heap_off + offset_from_heap + (i * one_elm_size);
            binary_read_recursive_Point(meta, binary_data, tmp_obj, item_offset);
            array_value.push(tmp_obj);
        }
        js_obj.cells = array_value;
    }
    
    return js_obj;
}

/**
 * Serializes a GridCells object into a binary PDU.
 * @param { GridCells } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_GridCells(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_GridCells(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_GridCells(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: header, type: std_msgs/Header (struct)

    {
        binary_write_recursive_Header(parent_off + 0, bw_container, allocator, js_obj.header);
    }
    
    // member: cell_width, type: float32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("float32", js_obj.cell_width, 4);
        allocator.add(bin, parent_off + 136);
    }
    
    // member: cell_height, type: float32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("float32", js_obj.cell_height, 4);
        allocator.add(bin, parent_off + 140);
    }
    
    // member: cells, type: geometry_msgs/Point (struct)

    { // varray
        const offset_from_heap = bw_container.heap_allocator.size();
        const array_size = js_obj.cells.length;
        const one_elm_size = 24;
        const array_base_offset = bw_container.heap_allocator.add(new ArrayBuffer(one_elm_size * array_size));
        for (let item_index = 0; item_index < array_size; item_index++) {
            const item_offset = array_base_offset + (item_index * one_elm_size);
            binary_write_recursive_Point(item_offset, bw_container, bw_container.heap_allocator, js_obj.cells[item_index]);
        }

        const ref_buffer = new ArrayBuffer(8);
        const ref_view = new DataView(ref_buffer);
        ref_view.setInt32(0, array_size, littleEndian); // array_size
        ref_view.setInt32(4, offset_from_heap, littleEndian);
        allocator.add(ref_buffer, parent_off + 144);
    }
    
}

File: pdu/csharp_v2/nav_msgs/pdu_conv_GridCells.cs
using System;
using System.Collections.Generic;

using Hakoniwa.Pdu.CSharpV2;
using Hakoniwa.Pdu.CSharpV2.nav_msgs;
using Hakoniwa.Pdu.CSharpV2.std_msgs;
using Hakoniwa.Pdu.CSharpV2.geometry_msgs;

namespace Hakoniwa.Pdu.CSharpV2.nav_msgs
{
    public static class GridCellsConverter
    {
        public static GridCells PduToMsg(byte[] binaryData)
        {
            var obj = new GridCells();
            var meta = PduMetaData.Parse(binaryData);
            BinaryReadRecursive(meta, binaryData, obj, PduMetaData.PduMetaDataSize);
            return obj;
        }

        public static byte[] MsgToPdu(GridCells obj)
        {
            var baseAllocator = new DynamicAllocator();
            var writer = new BinaryWriterContainer();
            BinaryWriteRecursive(0, writer, baseAllocator, obj);
            return PduRuntime.BuildPdu(baseAllocator, writer);
        }

        public static void BinaryReadRecursive(PduMetaData meta, byte[] binaryData, GridCells obj, int baseOff)
        {
            obj.header = new Header();
            HeaderConverter.BinaryReadRecursive(meta, binaryData, obj.header, baseOff + 0);
            obj.cell_width = PduRuntime.ReadFloat32(binaryData, baseOff + 136);
            obj.cell_height = PduRuntime.ReadFloat32(binaryData, baseOff + 140);
            obj.cells = new List<Point>();
            {
                var arraySize = PduRuntime.ReadInt32(binaryData, baseOff + 144);
                var offsetFromHeap = PduRuntime.ReadInt32(binaryData, baseOff + 144 + 4);
                for (var i = 0; i < arraySize; i++) {
                    var tmp = new Point();
                    PointConverter.BinaryReadRecursive(meta, binaryData, tmp, meta.HeapOff + offsetFromHeap + (i * 24));
                    obj.cells.Add(tmp);
                }
            }
        }

        public static void BinaryWriteRecursive(int parentOff, BinaryWriterContainer writer, DynamicAllocator allocator, GridCells obj)
        {
            HeaderConverter.BinaryWriteRecursive(parentOff + 0, writer, allocator, obj.header);
            allocator.Add(PduRuntime.GetBinaryForFloat32(obj.cell_width), parentOff + 136);
            allocator.Add(PduRuntime.GetBinaryForFloat32(obj.cell_height), parentOff + 140);
            {
                var offsetFromHeap = writer.HeapAllocator.Size();
                var arraySize = obj.cells.Count;
                allocator.Add(PduRuntime.GetBinaryForInt32(arraySize), parentOff + 144);
                allocator.Add(PduRuntime.GetBinaryForInt32(offsetFromHeap), parentOff + 144 + 4);
                writer.HeapAllocator.Add(new byte[arraySize * 24], offsetFromHeap);
                for (var i = 0; i < arraySize; i++) {
                    PointConverter.BinaryWriteRecursive(offsetFromHeap + (i * 24), writer, writer.HeapAllocator, obj.cells[i]);
                }
            }
        }
    }
}
